File: 202.java
class Solution {
    public boolean isHappy(int num) {
        int fast = num, slow = num;
        do {
            fast = getNext(getNext(fast));
            slow = getNext(slow);
        } while (fast != slow);
        return fast == 1;
    }

    private static int getNext(int num) {
        int sum = 0;
        int i;
        while (num > 0) {
            i = num % 10;
            sum += i*i;
            num /= 10;
        }
        return sum;
    }
}

File: 202.py
class Solution(object):
    def isHappy(self, n):
        """
        :type n: int
        :rtype: bool
        """
        res_set = []
        while n:
            if n == 1:
                return True
            if n not in res_set:
                res = sum(int(x)**2 for x in str(n))
                res_set.append(n)
                n = res
            else:
                return False

            